src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -28,18 +28,14 @@
 #include "memory/allocation.hpp"
 
 template<
+    typename TABLE_IMPL,
     typename K, typename V,
-    // xlC does not compile this:
-    // http://stackoverflow.com/questions/8532961/template-argument-of-type-that-is-defined-by-inner-typedef-from-other-template-c
-    //typename ResourceHashtableFns<K>::hash_fn   HASH   = primitive_hash<K>,
-    //typename ResourceHashtableFns<K>::equals_fn EQUALS = primitive_equals<K>,
-    unsigned (*HASH)  (K const&)           = primitive_hash<K>,
-    bool     (*EQUALS)(K const&, K const&) = primitive_equals<K>,
-    unsigned SIZE = 256,
-    ResourceObj::allocation_type ALLOC_TYPE = ResourceObj::RESOURCE_AREA,
-    MEMFLAGS MEM_TYPE = mtInternal
+    unsigned (*HASH)  (K const&),
+    bool     (*EQUALS)(K const&, K const&),
+    ResourceObj::allocation_type ALLOC_TYPE,
+    MEMFLAGS MEM_TYPE
     >
-class ResourceHashtable : public ResourceObj {
+class ResourceHashtableBase : public ResourceObj {
```

Comment:
Rather than a CRTP base class, I think it might be simpler to have a base class that has a type template parameter that provides the sizing/resizing policy. That type might be used either to specify the type of a new member or even a further base class (to benefit from EBO in the size-is-constant case). The derived class constructor would call the base class constructor with a policy object as an argument.

Reply:
Per Kim's suggestion, I moved the storage management code to two base classes: FixedResourceHashtableStorage and ResizeableResourceHashtableStorage. 

Now the `ResourceHashtable::_table[]` is in-line allocated (same as as before this PR). I checked with gcc and it generates identical code as before this PR.

Reply:
It's kind of unfortunate / confusing that ResourceHashtableBase has two "incompatible" constructors as a result of factoring out the storage, but otherwise I like this.  One consequence is that the two derived types actually need to be classes with different constructors, and can't just be type aliases of the base.